src/app/store/observeMiddleware.ts

```diff
@@ -0,0 +1,100 @@
+import { infoLog } from 'app/utils/log';
+import { Dispatch, isAction, Middleware, MiddlewareAPI } from 'redux';
+import { createAction, createCustomAction, isActionOf } from 'typesafe-actions';
+import { RootState } from './types';
+
+export interface StoreObserver<T> {
+  /**
+   * A unique string for the observer, this must be globally unique.
+   */
+  id: string;
+  /**
+   * A custom equality function. It not provided Objest.is will be used.
+   */
+  equals?: (a: unknown, b: unknown) => boolean;
+  /**
+   * Function to create "something" which will be used to determine if the side effect should run.
+   * deepEqual is used for the comparison, so use primitives, shallow objects or arrays to keep
+   * performance reasonable.
+   */
+  getObserved: (rootState: RootState) => T;
+  /**
+   * Runs the side effect providing both the previous and current version of the derrived state.
+   */
+  sideEffect: (states: { previous: T; current: T }) => void;
+}
+
+/**
+ * This is needed as the typesafe actions library throws a validation error for the observe action
+ * when created with `createCustomAction`.
+ * See https://github.com/DestinyItemManager/DIM/pull/10195/files#r1438195519
+ */
+function isObserveAction(action: unknown): action is ReturnType<typeof observe> {
+  return (
+    typeof action === 'object' &&
+    action !== null &&
+    'type' in action &&
+    action.type === 'observer/OBSERVE'
+  );
+}
+
+// Need to user a higher order function to get the correct typings and inference, it will now correctly
+// type the value of T based on what the getObserved returns
+export const observe = <T>(storeObserver: StoreObserver<T>) =>
+  createCustomAction('observer/OBSERVE', (storeObserver: StoreObserver<T>) => ({ storeObserver }))(
+    storeObserver,
+  );
+export const unobserve = createAction('observer/UNOBSERVE')<string>();
+export const clearObservers = createAction('observer/CLEAR_OBSERVERS')();
+
+export function observerMiddleware<D extends Dispatch>(
+  api: MiddlewareAPI<D, RootState>,
+): ReturnType<Middleware> {
+  const observers = new Map<string, StoreObserver<unknown>>();
+  return (next) => (action) => {
+    // Taken from the redux listener middleware, apparently some actions may not
+    // be action objects, https://github.com/reduxjs/redux-toolkit/blob/0d1f7101e83865714cb512c850bc53ffaee2d5e5/packages/toolkit/src/listenerMiddleware/index.ts#L438-L440
+    if (!isAction(action)) {
+      return next(action);
+    }
+
+    if (isObserveAction(action)) {
+      observers.set(action.storeObserver.id, action.storeObserver);
```

Comment:
I wonder if we should throw here if the key already exists to ensure that we never get observers with clashing id's. It might be safer and we can add a "replace" if we think it is absolutely necessary.